Introduce UI Element which supports an internal popup dialog

ihhub opened the pull request:
- Kingdom Income window was changed to have `Kingdom Income` title and `Kingdom Income per day` to be more obvious for a player
- Events on land and sea now show artifact and resources in one window ( close #1969 )
- Collected artifacts, treasure chests, shrines related dialogs contains right mouse clickable icons ( close #468 )
- Pyramid visit outcome show correct horizontal position of icons and all icons are right mouse clickable ( close #1726 )
- Icons in captured mines and sawmills dialogs are right mouse clickable
- Experience related window such as in Daemon Cave and Gazebo have right mouse clickable icons
- Add static methods for Font types
- Fix Artifact related drawings ( close #4616 )
- Code cleanup and adding more assertions

oleg-derevenetz commented on src/fheroes2/gui/ui_dialog.h:
```diff
@@ -20,9 +20,136 @@
+    public:
+        explicit ArtifactDialogElement( const Artifact & artifact );
+
+        ~ArtifactDialogElement() override = default;
```
I believe this is not necessary to explicitly override a virtual destructor of the base class by the default destructor here and in the other similar places below, but it's a matter of taste :)

ihhub replied:
I did this as clang-tidy as well as compilers were complaining about it.

oleg-derevenetz reviewed (approved):
Hi @ihhub LGTM, just one optional comment.

ihhub commented:
Hi @oleg-derevenetz , could you please take a look again? I expanded the changes to have a completed pull request. I removed a lot of old dialog based code and fixed some issues as well standardized Text types.

LeHerosInconnu commented:
Hello @ihhub,

> Hi @Branikolog and @LeHerosInconnu, I am asking you to help to test these changes as they impact a lot of dialogs.

I don't have too much time right now, I'll look at this towards the end of next week. :)

zenseii commented:
This looks very good! I only found some small points that could be discussed or fixed:

Fixed, in my opinion:

1. Right-clickable experience icon and gold icon when picking up a treasure chest.

2. Experience info English text - Instead of "go up a level" I would say "gain a level" just to stay coherent with the rest of the game.

3. Experience info English text - Add text saying that gaining a level let's you learn a new secondary skill or improve one (yes, at high levels you won't gain anything new when maxed out)

Discussable:

1. Pyramid/Shrine/Etc. - In case you cannot learn a spell, should the spell info still be possible to be found by right-clicking something?

2. Right-clickable resource in the lower right status window when you pick up something on the adventure map?

ihhub commented:
> This looks very good! I only found some small points that could be discussed or fixed:
> 
> Fixed, in my opinion:
> 
>     1. Right-clickable experience icon and gold icon when picking up a treasure chest.
> 
>     2. Experience info English text - Instead of "go up a level" I would say "gain a level" just to stay coherent with the rest of the game.
> 
>     3. Experience info English text - Add text saying that gaining a level lets you learn a new secondary skill or improve one (yes, at high levels you won't gain/improve any new secondary skills when maxed out.)
> 
> 
> Discussable:
> 
>     1. Pyramid/Shrine/Etc. - In case you cannot learn a spell, because of insufficient Wisdom skill, should the spell info still be possible to find by right-clicking something? I believe there's an issue/discussion already concerning this.
> 
>     2. Right-clickable resource in the lower right status window when you pick up something on the adventure map?
> 
> 
> Thank you for the work put into this PR!

Hi @zenseii , thanks for your review!
Regarding your comments:
1. It is not a part of the standard dialog introduced in this pull request. It will be fixed later.
2 and 3. The text appears to be the same as in the original game. I wrote in the exact same way. We can refine it later.

Discussable notes:
1. It can be done with `CustomImageDialogElement` class but as you remarked it is debatable.
2. This window is not a part of the standard window classes.

ihhub commented:
I am merging this pull request as it is to do not hold on any further. If we find something broken or needed an improvement we are going to fix it in a separate pull request.

LeHerosInconnu commented:
Hello @ihhub,

> Hello @ihhub,
> 
> > Hi @Branikolog and @LeHerosInconnu, I am asking you to help to test these changes as they impact a lot of dialogs.
> 
> I don't have too much time right now, I'll look at this towards the end of next week. :)

I spotted a few problems while testing this, I will report them in new issues. :)
